### ISSUE TITLE
Better roundtrip of purl() and spin()

### ISSUE THREAD

**Issue by krlmlr** [yihui/knitr#2014]
Occasionally I want to convert an `.Rmd` file to `.R` to execute it in the console until the first error. When I edit the `.R` file, the conversion back to `.Rmd` contains spurious differences, most notably empty lines.

Can we work towards ensuring that `spin(purl(spin(purl())))` is equivalent to `spin(purl())`, i.e. that the roundtrip is stable after the second iteration?

``` r
text <- "Text\n\n```{r chunk}\n1 + 2\n3 + 4\n```\n\nMore text\n```{r chunk2}\n5 + 6\n```\n"

writeLines(text, "test.Rmd")

library(knitr)

purl("test.Rmd", documentation = 2)
#> processing file: test.Rmd
#> output file: test.R
#> [1] "test.R"
spin("test.R", knit = FALSE)
#> [1] "test.Rmd"

test_1 <- readLines("test.Rmd")

purl("test.Rmd", documentation = 2)
#> processing file: test.Rmd
#> output file: test.R
#> [1] "test.R"
spin("test.R", knit = FALSE)
#> [1] "test.Rmd"

test_2 <- readLines("test.Rmd")

waldo::compare(test_1, test_2)
#> `old`: "Text" "" ""    "```{r chunk}" "1 + 2" "3 + 4" "```"    "" and 10 more...
#> `new`: "Text" "" "" "" "```{r chunk}" "1 + 2" "3 + 4" "```" "" "" ...
```

<sup>Created on 2021-06-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by krlmlr** [comment]
The roundtripping seems to improve if we use {styler} to style the `.R` and/or the `.Rmd` file.

**Comment by yihui** [comment]
Yes, that would be ideal. I guess it will break some tests, but I'm fine with it (the only trouble is that it may break tests in other packages, and I'll need to work with these package authors). Anyway, please feel free to send a pull request, and I'll see if it can be merged without too much trouble. Thanks!

**Comment by yihui** [comment]
Thanks @krlmlr — the reprex made this easy to pin down.

Fixed in 32b63a055472356df489cbc5b9a52b6353ea8347. The concrete bug was that `purl(documentation = 2)` writes `## ----label----` separators, and `spin()` was gluing the label straight onto the engine name, producing an invalid header like ```{rlabel}. On the next `purl()`, `rlabel` was read as an engine name and the chunk got mis-parsed (code commented out) — that's the bulk of the round-trip breakage in your example. `spin()` now emits ```{r label}.

One thing I did **not** change: the blank lines still accumulate slightly across repeated conversions. As you noted, running a styler over the output is the intended way to fully normalize whitespace; making `spin()`/`purl()` byte-stable on their own would be a much larger change. If the residual whitespace churn is a real problem for your workflow, feel free to open a separate focused issue.
